### PR TITLE
Make ProductInstances lazy so derivation works for recursive types

### PR DESCRIPTION
**Pull request by TimWSpence**

### CODE REVIEW

**Comment by TimWSpence** [comment]
Compilation hangs on that new test case if I remove the `=>` so it seems to work as expected

**Comment by joroKr21** [comment]
We need to think about whether we can break binary compatibility here - those classes are considered internal, but there is a lot of inlining happening here and I'm not sure.

**Comment by TimWSpence** [comment]
> We need to think about whether we can break binary compatibility here - those classes are considered internal, but there is a lot of inlining happening here and I'm not sure.

Mima seems to think that it will break bincompat. Not sure if there's anything we can do about that? I also can't figure out where the version is defined if we do want to bump the major version 😂

**Comment by joroKr21** [comment]
We definitely don't want to bump the major version 😄 - but those classes are in the `internals` package because they are not supposed to be used directly by users. So we could just ignore this package from MiMa checks. I'm just not 100% convinced that it wouldn't cause issues anyway because of inlining.

**Comment by milessabin** [comment]
LGTM apart from the major version bump!

**Comment by TimWSpence** [comment]
> We definitely don't want to bump the major version 😄 - but those classes are in the `internals` package because they are not supposed to be used directly by users. So we could just ignore this package from MiMa checks. I'm just not 100% convinced that it wouldn't cause issues anyway because of inlining.

Good question, I'm not sure of the impact of inlining.

Do we have any idea how many projects are currently depending on shapeless 3 and hence affected if we ever did break bincompat?

**Comment by TimWSpence** [comment]
> We definitely don't want to bump the major version 😄 - but those classes are in the `internals` package because they are not supposed to be used directly by users. So we could just ignore this package from MiMa checks. I'm just not 100% convinced that it wouldn't cause issues anyway because of inlining.

🤦 I hadn't realized that that version is the only thing distinguishing this from shapeless 2! I thought we were publishing a separately versioned shapeless 3 artifact but we're not

**Comment by joroKr21** [comment]
The organisation is also different. But no I'm not aware of the current usage of shapeless. I see that there are already some breaking changes we ignore in the `internals` package. On the other hand it's also possible to make the primary constructor take `() => Array[Any]`, add a secondary constructor which is eager and change the `ErasedProductInstances.apply` method. If we're not sure this would preserve binary compatibility.

**Comment by TimWSpence** [comment]
@joroKr21 I just set up a test project with a single file (https://gist.github.com/TimWSpence/7083c57567789b717d56cfc4514b1289) and compiled it via sbt using shapeless 3.0.3. I then ran it manually with `java` with an explicit classpath. Swapping shapeless 3.0.3 for a locally published snapshot gives
```
Exception in thread "main" java.lang.NoSuchMethodError: 'void shapeless3.deriving.internals.ErasedProductInstances1.<init>(scala.deriving.Mirror$Product, java.lang.Object)'
        at Main$package$.run(Main.scala:52)
        at run.main(Main.scala:45)

```

😭 So yes, I think inlining does cause bin-compat problems

**Comment by joroKr21** [comment]
Thanks for checking @TimWSpence ! Yeah in that case we should go for the binary compatible solution with a secondary constructor.

File: modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala [hunk 1]
```diff
@@ -41,7 +41,15 @@ private[shapeless3] abstract class ErasedProductInstances[K, FT] extends ErasedI
   def erasedProject(x0: Any)(p: Int)(f: (Any, Any) => Any): Any
 }
 
-private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirror.Product, i: Any) extends ErasedProductInstances[K, FT] {
+private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirror.Product, _i: () => Any) extends ErasedProductInstances[K, FT] {
+
+  //Preserved for bincompat reasons. DO NOT USE as it will lead to stack overflows when deriving
+  //instances for recursive types
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala:47]: Can we make that the message of `@deprecated`?

File: modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala [hunk 2]
```diff
@@ -99,7 +107,20 @@ private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirro
...
-private[shapeless3] final class ErasedProductInstancesN[K, FT](val mirror: Mirror.Product, is: Array[Any]) extends ErasedProductInstances[K, FT] {
+object ErasedProductInstances1 {
+  def apply[K, FT](mirror: Mirror.Product, i: => Any): ErasedProductInstances1[K, FT] =
+    new ErasedProductInstances1(mirror, () => i)
+}
+
+private[shapeless3] final class ErasedProductInstancesN[K, FT](val mirror: Mirror.Product, _is: () => Array[Any]) extends ErasedProductInstances[K, FT] {
+
+  //Preserved for bincompat reasons. DO NOT USE as it will lead to stack overflows when deriving
+  //instances for recursive types
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala:118]: Same here

File: modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala [hunk 3]
```diff
@@ -41,7 +41,15 @@ private[shapeless3] abstract class ErasedProductInstances[K, FT] extends ErasedI
   def erasedProject(x0: Any)(p: Int)(f: (Any, Any) => Any): Any
 }
 
-private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirror.Product, i: Any) extends ErasedProductInstances[K, FT] {
+private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirror.Product, _i: () => Any) extends ErasedProductInstances[K, FT] {
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala:44]: Just for consistency
```suggestion
private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirror.Product, i0: () => Any) extends ErasedProductInstances[K, FT] {
```

File: modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala [hunk 4]
```diff
@@ -99,7 +107,20 @@ private[shapeless3] final class ErasedProductInstances1[K, FT](val mirror: Mirro
     f(i, toProduct(x0).productElement(0))
 }
 
-private[shapeless3] final class ErasedProductInstancesN[K, FT](val mirror: Mirror.Product, is: Array[Any]) extends ErasedProductInstances[K, FT] {
+object ErasedProductInstances1 {
+  def apply[K, FT](mirror: Mirror.Product, i: => Any): ErasedProductInstances1[K, FT] =
+    new ErasedProductInstances1(mirror, () => i)
+}
+
+private[shapeless3] final class ErasedProductInstancesN[K, FT](val mirror: Mirror.Product, _is: () => Array[Any]) extends ErasedProductInstances[K, FT] {
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala:115]: here too:
```suggestion
private[shapeless3] final class ErasedProductInstancesN[K, FT](val mirror: Mirror.Product, is0: () => Array[Any]) extends ErasedProductInstances[K, FT] {
```

**Review by joroKr21 — CHANGES_REQUESTED** [review]
Other than that LGTM - awesome work 👍

**Review by joroKr21 — APPROVED** [review]
Nice!

**Comment by joroKr21** [comment]
Should we merge #44 so we can release this?

**Comment by TimWSpence** [comment]
> Should we merge #44 so we can release this?

Fine by me! So have we never released for 3.1.0?

**Comment by joroKr21** [comment]
No we haven't yet

**Pull request merged**